xarray/core/variable.py

```diff
@@ -376,29 +384,121 @@ def _item_key_to_tuple(self, key):
+                indexer.append(k)
+            else:
+                k = np.asarray(k)
+                if k.ndim > 1:
+                    raise IndexError("Unlabelled multi-dimensional array "
+                                     "cannot be used for indexing.")
+                indexer.append(k if k.dtype.kind != 'b' else k.nonzero()[0])
+        return dims, OuterIndexer(indexer)
+
+    def nonzero(self):
+        """ Equivalent numpy's nonzero but returns a tuple of Varibles. """
+        # TODO we should replace dask's native nonzero
+        # after https://github.com/dask/dask/issues/1076 is implemented.
+        nonzeros = np.nonzero(self.data)
+        return tuple([as_variable(nz, name=dim) for nz, dim
+                      in zip(nonzeros, self.dims)])
+
+    def _broadcast_indexes_advanced(self, key):
+        variables = []
+
+        for dim, value in zip(self.dims, key):
+            if isinstance(value, slice):
+                value = np.arange(*value.indices(self.sizes[dim]))
+
+            try:
+                variable = as_variable(value, name=dim)
+            except MissingDimensionsError:  # change to better exception
+                raise IndexError("Unlabelled multi-dimensional array "
+                                 "cannot be used for indexing.")
+
+            if variable.dtype.kind == 'b':  # boolean indexing case
+                if variable.ndim > 1:
+                    raise IndexError("{}-dimensional boolean indexing is "
+                                     "not supported. ".format(variable.ndim))
+                variables.extend(list(variable.nonzero()))
+            else:
+                variables.append(variable)
+        variables = _broadcast_compat_variables(*variables)
+        dims = variables[0].dims  # all variables have the same dims
+        # overwrite if there is integers
+        key = VectorizedIndexer(k if isinstance(k, integer_types)
+                                else variable.data
+                                for variable, k in zip(variables, key))
+        return dims, key
+
     def __getitem__(self, key):
         """Return a new Array object whose contents are consistent with
         getting the provided key from the underlying data.
 
-        NB. __getitem__ and __setitem__ implement "orthogonal indexing" like
-        netCDF4-python, where the key can only include integers, slices
-        (including `Ellipsis`) and 1d arrays, each of which are applied
-        orthogonally along their respective dimensions.
-
-        The difference does not matter in most cases unless you are using
-        numpy's "fancy indexing," which can otherwise result in data arrays
-        whose shapes is inconsistent (or just uninterpretable with) with the
-        variable's dimensions.
+        NB. __getitem__ and __setitem__ implement "diagonal indexing" like
+        np.ndarray.
 
-        If you really want to do indexing like `x[x > 0]`, manipulate the numpy
```

Comment:
I think we still need this around, until we support boolean indexing with multi-dimensional indexers.